core/src/main/java/org/apache/brooklyn/core/objs/BasicSpecParameter.java

```diff
@@ -404,7 +431,126 @@ public static void addParameters(AbstractBrooklynObjectSpec<?, ?> spec, List<? e
+            super(Preconditions.checkNotNull(label!=null ? label : config.getName(), "label or config name must be set"), 
+                    pinned==null ? true : pinned, config, sensor);
+            this.hasType = hasType;
+            this.hasLabelSet = label!=null;
+            this.hasPinnedSet = pinned!=null;
+            this.hasDefaultValue = hasDefaultValue;
+            this.hasConstraints = hasConstraints;
+            this.hasRuntimeInheritance = hasRuntimeInheritance;
+            this.hasTypeInheritance = hasTypeInheritance;
+        }
+
+        /** used if yaml specifies a parameter, but possibly incomplete, and a spec supertype has a parameter */
+        SpecParameter<?> resolveWithAncestor(SpecParameter<?> ancestor) {
+            if (ancestor==null) return new BasicSpecParameter<>(getLabel(), isPinned(), getConfigKey(), getSensor());
+
+            return new BasicSpecParameter<>(
+                    hasLabelSet ? getLabel() : ancestor.getLabel(), 
+                            hasPinnedSet ? isPinned() : ancestor.isPinned(), 
+                                    resolveWithAncestorConfigKey(ancestor.getConfigKey()), 
+                                    hasType ? getSensor() : ancestor.getSensor());
+        }
+
+        /** as {@link #resolveWithAncestor(SpecParameter)} but where the param redefines/extends a config key coming from a java supertype, rather than a parameter */
+        // TODO not used yet; see calls to the other resolveWithAncestor method,
+        // and see BrooklynComponentTemplateResolver.findAllConfigKeys;
+        // also note whilst it is easiest to do this here, logically it is messy,
+        // and arguably it should be done when converting the spec to an instance
+        SpecParameter<?> resolveWithAncestor(ConfigKey<?> ancestor) {
+            if (ancestor==null) return new BasicSpecParameter<>(getLabel(), isPinned(), getConfigKey(), getSensor());
+
+            // TODO probably want to do this (but it could get expensive!)
+            //          Set<Class<?>> types = MutableSet.<Class<?>>builder()
+            //                  .add(spec.getImplementation())
+            //                  .add(spec.getType())
+            //                  .addAll(spec.getAdditionalInterfaces())
+            //                  .remove(null)
+            //                  .build();
+            //          // order above is important, respected below to take the first one defined 
+            //          MutableMap<String, ConfigKey<?>> result = MutableMap.copyOf(FlagUtils.findAllConfigKeys(null, types));
+
+            return new BasicSpecParameter<>(
+                    getLabel(), 
+                    isPinned(), 
+                    resolveWithAncestorConfigKey(ancestor),
+                    // TODO port sensor will be lost (see messy code above which sets the port sensor)
```

Comment:
no, we've never done proper combining of sensors when combinining a config definition with an ancestor.  and btw this method isn't really used yet, at least this code path isn't, it's in the TODO code path at https://github.com/apache/brooklyn-server/pull/462/files/a777b36e16688ccf9ef53ad6c1d392c8ef0f85bc#diff-94206c636ff5bbab4c4251c3316c129e